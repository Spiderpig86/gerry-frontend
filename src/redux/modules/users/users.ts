// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
import { User } from "models";
import { authenticate, checkIfAuthenticated } from "components";

const AUTH_USER = 'AUTH_USER'
const UNAUTH_USER = 'UNAUTH_USER'
const FETCHING_USER = 'FETCHING_USER'
const FETCHING_USER_FAILURE = 'FETCHING_USER_FAILURE'
const FETCHING_USER_SUCCESS = 'FETCHING_USER_SUCCESS'

const authUser = (uid: string) => {
  return {
    type: AUTH_USER,
    uid,
  }
}

export const unauthUser = () => {
  return {
    type: UNAUTH_USER,
  }
}

const fetchingUser = () => {
  return {
    type: FETCHING_USER,
  }
}

const fetchingUserFailure = (error: string) => {
  return {
    type: FETCHING_USER_FAILURE,
    error: 'Error fetching user.',
  }
}

const fetchingUserSuccess  = (uid: string, user: User, timestamp: Date) => {
  return {
    type: FETCHING_USER_SUCCESS,
    uid,
    user,
    timestamp,
  }
}

export const fetchAndHandleAuthentication = (history: any) => {
  return (dispatch: any) => {
    dispatch(fetchingUser());
    authenticate().then((user: any) => {
      dispatch(fetchingUserSuccess(user.uid, user, new Date()));
      dispatch(authUser(user.uid));
      history.push("/dashboard");
    }).catch((error) => dispatch(fetchingUserFailure(error)));
  };
}

const initialUserState: User = {
    uid: "",
    name: "",
    lastUpdated: new Date()
}

export const user = (state = initialUserState, action: any) => {
  switch (action.type) {
    case FETCHING_USER_SUCCESS :
      return {
        ...state,
        uid: action.uid,
        name: action.user.name,
        lastUpdated: action.timestamp
      }
    default :
      return state
  }
}

const initialState = {
  isFetching: false,
  error: "",
  isAuthed: false,
  authedId: "",
}

export const users = (state = initialState, action: any) => {
  switch (action.type) {
    case AUTH_USER :
      return {
        ...state,
        isAuthed: true,
        authedId: action.uid,
      }
    case UNAUTH_USER :
      return {
        ...state,
        isAuthed: false,
        authedId: "",
      }
    case FETCHING_USER:
      return {
        ...state,
        isFetching: true,
      }
    case FETCHING_USER_FAILURE:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    case FETCHING_USER_SUCCESS:
      return action.user === null
        ? {
          ...state,
          isFetching: false,
          error: "",
        }
        : {
          ...state,
          isFetching: false,
          error: "",
          [action.uid]: user(state[action.uid], action),
        }
    default :
      return state
  }
}