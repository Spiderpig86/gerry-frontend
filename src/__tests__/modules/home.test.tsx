// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
import React from "react";
import renderer from "react-test-renderer";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { users } from "../../redux/modules/users/users"
import { Home } from "modules";

describe("Home component tests", () => {
    it("Check Home Component Render", () => {
        const store = createStore(users, applyMiddleware(thunk));
      const component = renderer.create(<Provider store={store}><Home history={[]} /></Provider>).toJSON();
      expect(component).toMatchSnapshot();
    });
});