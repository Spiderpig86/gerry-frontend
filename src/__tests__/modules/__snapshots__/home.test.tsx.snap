// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component tests Check Home Component Render 1`] = `
<div
  className="container-fluid home"
>
  <div
    className="row"
  >
    <div
      className="hero-container"
    >
      <button
        className="btn btn-white hero-sign-in"
        onClick={[Function]}
      >
        Sign in
        <i
          className="glyphicon glyphicon-menu-right"
        />
      </button>
      
    </div>
  </div>
</div>
`;
